### PR TITLE
Update to JupyterLab 3.4 and fix toolbar

### LINKED ISSUE #31: Missing file browser toolbar items with JupyterLab 3.4.0
## Description

The file browser toolbar is missing a couple of items with JupyterLab 3.4.0.

**With JupyterLab 3.3.4**

image

**With JupyterLab 3.4.0**

image

## Reproduce

- `pip install jupyterlab>=3.4.0 jupyterlab-filesystem-access`
- Open JupyterLab

## Context

This is most likely related to https://github.com/jupyterlab/jupyterlab/pull/12281

- Python package version: 0.4.0

- Extension version: 0.4.0

- Operating System and its version:  Ubuntu 20.04 - Chromium 101

### PR DESCRIPTION
**Pull request by martinRenou**
Fix #31

### CODE REVIEW

**Comment by martinRenou** [comment]
This is currently not working properly, the issue with this code it that it "steals" the toolbar buttons from the main filebrowser. The toolbar buttons don't show up anymore on the main filebrowser, they show up only on our custom one.

toolbar

File: src/index.ts [hunk 1]
```diff
@@ -75,6 +76,23 @@ const plugin: JupyterFrontEndPlugin<void> = {
...
+      (browser: FileBrowser) => openDirectoryButton
+    );
+
+    setToolbar(
+      widget,
+      createToolbarFactory(
+        toolbarRegistry,
+        settingRegistry!,
+        'FileBrowser',
+        '@jupyterlab/filebrowser-extension:widget',
```
Comment (jtpio) [inline src/index.ts:91]: Maybe we should define our own toolbar items in the settings and change this to the following?

```suggestion
        'jupyterlab-filesystem-access:plugin',
```

Comment (martinRenou) [reply src/index.ts:91]: Given that we want to remove the "Open Launcher" button I guess it makes sense to do that indeed.

**Comment by martinRenou** [comment]
I pushed another commit that seems to work. I couldn't make it work with a custom config without breaking the main Filebrowser.

**Comment by fcollonval** [comment]
The toolbar generator from settings schema is registering a transformation of the settings to build them by gathering definition from multiple extensions. When using the same helper on the same settings, it fails as it is forbidden to register multiple transform operators on the same plugin.

This definitely needs to be fixed. But as mentioned above, as some buttons are not valid for this file browser, it will be better to use this extension settings.

> Problematic call:

https://github.com/jupyterlab/jupyterlab/blob/a41b9816fdedb936be1589dbc98363f62f5f4574/packages/apputils/src/toolbar/factory.ts#L109

**Comment by fcollonval** [comment]
@martinRenou I pushed a patch to use the settings to customize the toolbar. It is independent of the standard filebrowser toolbar definition, hence the additional code (+ the need to wait for an upstream fix in jupyterlab/jupyterlab#12556)

**Comment by martinRenou** [comment]
Thanks!

**Comment by martinRenou** [comment]
I'll give this another try tomorrow :)

**Comment by psychemedia** [comment]
Is this still an open issue? I've lost track of what's working where across JL3/4 "standard" and lite (and where I had things that were woking but unpinned they're starting to break in ways that are hard to track as I try to update them for the new academic year!)

**Comment by martinRenou** [comment]
Yes this is still needed.

This extension does not work with JupyterLab 4.

**Comment by psychemedia** [comment]
Aside from JL4, is this fixed in JL3 too? Or is the extension deprecated now for JLv3 latest?

**Comment by martinRenou** [comment]
Sorry I did not mean to close this. It is not fixed for JupyterLab 4 or 3.

**Comment by jtpio** [comment]
There is an issue to update to JupyterLab 4: https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access/issues/54

**Comment by jtpio** [comment]
@fcollonval I'm trying to pick this up as part of https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access/pull/55.

What would be to define toolbar items via the setting system for this custom drive, so it doesn't conflict with the default filebrowser? This custom drive uses `createFileBrowser` to instantiate a new file browser, should the configuration still reference `FileBrowser` like the following?

```json
"jupyter.lab.toolbars": {
  "FileBrowser": [
    { "name": "open-folder", "rank": 0 }
  ]
},
```

**Comment by jtpio** [comment]
Answering my own question: this is possible using a combination of `setToolbar` and `createToolbarFactory`.

I'll check if there are mentions to that in the JupyterLab docs and post them here. Or open an issue in JupyterLab otherwise.

File: src/index.ts [hunk 2]
```diff
@@ -60,25 +60,151 @@ const plugin: JupyterFrontEndPlugin<void> = {
...
 
-    const openDirectoryButton = new ToolbarButton({
-      icon: folderIcon,
-      onClick: async () => {
-        const directoryHandle = await window.showDirectoryPicker();
-
-        if (directoryHandle) {
-          drive.rootHandle = directoryHandle;
+    toolbarRegistry.registerFactory(
+      'FileBrowser', // Factory name
```
Comment (fcollonval) [inline src/index.ts:64]: @jtpio the `FileBrowser` entry in the toolbars section is to match this name. If here you define a different name, you will be your own master for items that goes in this filebrowser toolbar. But it also means that for example you will not get the `Add Launcher` button as it targets the default `FileBrowser` namespace.

Does it clarify your question?

Comment (jtpio) [reply src/index.ts:64]: yes thanks! I think what we need here is what is described in https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html#generic-widget-with-toolbar

Comment (jtpio) [reply src/index.ts:64]: I'll push some changes in https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access/pull/55.

**Comment by jtpio** [comment]
> I'll check if there are mentions to that in the JupyterLab docs and post them here

This is the relevant bit: https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html#generic-widget-with-toolbar

**Pull request closed without merging**